SWIPs/swip-multiple_payment_processing_support.md

```diff
@@ -27,53 +27,162 @@ This SWIP is part of a series of SWIPs (but can be implemented on its own). To s
 
-When it becomes possible for nodes to set their preference for a payment module, developers will be incentivized to implement payment modules on Swarm as it will be easy for users to choose to pay with this module. It will also seamlessly enable multicurrency support and foster interoperation across blockchains without forcing participants to be tied to a single Blockchain or settlement technology.
+When it becomes possible for nodes to set their preference for a payment module, developers will be incentivized to implement payment modules on Swarm as it will be easy for users to choose to pay with this module. It will also seamlessly enable multicurrency support and foster interoperation across Blockchains without forcing participants to be tied to a single Blockchain or settlement technology.
 
 ## Specification
 <!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for the current Swarm platform and future client implementations.-->
 
-At a high level a payment module is responsible for:
+### Generic payment module
+At high level a payment module is responsible for sending a payment to a recipient (a peer providing storage services) for services provided. How payments are actually performed will depend on the specific implementation of the payment module. The payment module is then responsible for:
 
-* Accepting an amount (in honey) and a recipient.
-* Resolving the conversion from honey to money (currently by querying the agreed-upon price oracle with the recipient.)
+* Resolving the conversion from honey to money.
 * Ensuring that the user can engage in SWAP accounting for the chosen payment module before payment is due.
-* Sending the recipient a payment.
-* Returning true when the payment was successful.
-* Returning false when the payment was not successful.
-* Verifying any on-chain conditions and making sure that they don't change within a reasonable time window after the transaction happens (due to block reorganizations / payment-module specific conditions)
-* Referencing a type, version and base currency 
+* Sending a payment to the recipient.
+* Sending a message to the accounting module upon receipt of payment
+* Ensuring that the user can send payments with the chosen payment module before payment is due.
+* Handling any potential errors.
 * Optionally exposing other methods such as querying balances, topping up balances or sending payments (outside of Swarm). 
 
-Nodes can specify their preference for both payment module, as well as price oracle in a list in a configuration file. The preferences are normalized and weighed.
-For any preference list, the chosen option will be the option which has the highest cumulative preference. The preference list has three dimensions, which will be resolved from high to low:
+### Allowing multiple payment modules
```

Comment:
Would it make sense to stop describing this process as something that happens in the existing handshake? The truth is that we don't need this negotiation to happen when the connection with the other peer occurs, we can delay the negotiation to a later time, since the payment module will be needed when the node needs to issue a payment. Pushing this idea to the extreme it could be possible to delay it until the payment threshold is crossed.

Reply:
I think it is a good idea to do this negotiation at the point when the payment threshold is triggered as it will help to reduce the load when nodes build up a connection with the network. The disadvantage is that it will be unclear in which currency you are going to be paid out by the nodes who have never done a payment to you before, but that it IMO not a very big disadvantage, as the payment treshhold will not be high enough for this to matter.